Enable unbundling dependencies and linking to the system libraries instead.

perfinion opened the pull request:
This series adds a framework to be able to unbundle the dependencies from tensorflow. Previously, bazel rebuilds every single dep from scratch. This allows distro packages to disable bundling on a per-dep basis and link with the libraries that already exist on the system.

For more information why deps should be unbundled, see these links: https://wiki.gentoo.org/wiki/Why_not_bundle_dependencies
https://fedoraproject.org/wiki/Bundled_Libraries

The deps I have unbundled so far work on my Gentoo machine. Thanks to dennisjenkins@google.com for testing this on Gentoo as well. There are still more that need to be unbundled but this covers enough to be useful already. This has not been tested on any other distro than Gentoo but I don't forsee any big incompatibilities.

The libs that are unbundled are configured by setting `build --action_env TF_SYSTEM_LIBS="comma,sep,list"`
eg:
```
build --action_env TF_SYSTEM_LIBS="com_googlesource_code_re2,nasm,jpeg,png_archive,org_sqlite,gif_archive,six_archive,astor_archive,termcolor_archive,pcre,swig,curl,grpc,lmdb,zlib_archive,snappy,flatbuffers,cython,jemalloc"
```

Once configured, the tarball for the dep will not be downloaded or extracted, and the BUILD file will be swapped to the system_build_file instead which contains different rules to compile and link against the system package.
There are also macros if_system_lib(name, a, b) and if_not_system_lib(name, a, b) to configure other things in the build.

DO NOT MERGE THIS YET. I'm opening this for comments are review here before its ready for merging.

perfinion committed:
third_party/repo: add TF_SYSTEM_LIBS to unbundle deps

Signed-off-by: Jason Zaman <jason@perfinion.com>

perfinion committed:
workspace: use TF_SYSTEM_LIBS to link to system libraries

Signed-off-by: Jason Zaman <jason@perfinion.com>

perfinion committed:
third_party: Add astor systemlib

Signed-off-by: Jason Zaman <jason@perfinion.com>

perfinion committed:
third_party: Add cython system dep

Signed-off-by: Jason Zaman <jason@perfinion.com>

perfinion committed:
systemlibs: Add syslibs_configure repository_rule to generate if_system_lib macros

Signed-off-by: Jason Zaman <jason@perfinion.com>

perfinion committed:
workspace: configure system libs

Signed-off-by: Jason Zaman <jason@perfinion.com>

perfinion committed:
pip_package: exclude grpc licenses when building system library

The grpc license files are deeper in the heirarchy so are hard to nop
out in the system version of the BUILD file.

Signed-off-by: Jason Zaman <jason@perfinion.com>

perfinion commented:
@angersson so you can make sure this doesn't conflict with your changes.
@meteorcloudy can you take a look at the bazel bits?

angersson commented:
@jart Can you have a look at this as well?

The gist of this looks good to me, but I'm not certain of the implications for other systems, etc. I'm working on updating my internal proposal to better work with this.

perfinion committed:
third_party: unbundle jsoncpp dep

The jsoncpp headers are included with a different path so we have to
symlink them so the are in the dir structure that is expected.

Signed-off-by: Jason Zaman <jason@perfinion.com>

meteorcloudy commented on third_party/systemlibs/syslibs_configure.bzl:
```diff
@@ -0,0 +1,149 @@
+  fail("\n%sSystem Library Configuration Error:%s %s\n" % (red, no_color, msg))
+
+
+def _enable_syslibs(repository_ctx):
```
Since this is Linux specific feature, can you disable it on Windows?
You can check if it's building on Windows with a function like:
https://github.com/tensorflow/tensorflow/blob/fb3ce0467766a8e91f1da0ad7ada7c24fde7a73a/third_party/py/python_configure.bzl#L31-L36

perfinion replied:
Good call, I added this

meteorcloudy reviewed (dismissed):
Just a comment about Windows.
LGTM from Bazel side, thanks for the great work!

perfinion commented:
I pushed an update:

* disabled the syslibs_configure stuff on windows.
* fixed the search path in cython.BUILD.
* rebased on to latest master.
* master had bumped jsoncpp too so I added a commit to unbundle that as well. 
* grpc.BUILD added targets for the secure libs too.

The header file paths for jsoncpp are different from the system paths so I had to symlink them. bazel complains when I tried doing includes=["/usr/include/jsoncpp"]. Maybe in a future bazel version we can add some configs to allow non- hermetic builds using the system headers or searching pkg-config or something.

I've run this patchset in the gentoo ebuild both bundled and unbundled and its working so far :)

msarahan commented:
Thank you for working on this!  I've been trying to use this for Anaconda's packages, too.  I'm having some difficulty with the hardcoding of paths, such as with cython: https://github.com/tensorflow/tensorflow/pull/20284/files#diff-c8176cf1a58a40560176239b30ef7e52R8

Is there a good way to have that be more dynamic?  For example, all of our build paths are relative to an environment variable, ``$PREFIX``.  For you, this is probably ``/usr`` or maybe ``/usr/local``.  For us, it can be anything, and essentially never ``/usr``.  Perhaps we could use ``$PREFIX`` in the build files, and have it fall back to a default of ``/usr`` when it is not defined?

perfinion commented:
I have pushed sci-libs/tensorflow-1.9.0_rc1-r2 to the gentoo tree with these patches. the system-libs USE-flag enabled the system ones, disabling the flag uses the bundled stuff like before in case there are any issues. Hopefully I dont get too many bugs filed because of it :D

perfinion commented:
@msarahan yeah, the prefixing is something that I forsaw as problematic. Normal gentoo uses $ROOT=/, but we definitely do support both cross-compiling (eg $ROOT=/usr/arm-linux-gnu/) and also installing things as a whole alongside another distro/machine/whatever (eg $PREFIX=/home/jason/root/). Currently those cases are probably also broken. 

I was thinking to write some simple scripts like find-binary.sh and find-python.sh and change the genrule()'s to use those instead of directly cp'ing fixed paths. Once there are scripts then it would be quite easy to add extra logic per-distro or to search pkg-config for whatever flags. Not every distro puts binaries in the same place too so scripts would be more robust.

About $PREFIX specifically, bazel strips basically everything from the environment currently. There are issues filed with the bazel team and it will eventually more variables. In the mean time you'd need to add something like build --action_env=PREFIX.

This patch set is already getting pretty large so I think we should get this basic functionality merged in then later patchsets will be smaller and much easier to review :) Other than editing some paths in third_party/systemlibs/*.BUILD, has the rest mostly been working for you? If some of the core bits fail horribly for your case, that would probably warrant fixing first. You can look at the gentoo ebuilds for inspiration too but they aren't pretty yet: https://github.com/gentoo/gentoo/tree/master/sci-libs/tensorflow

msarahan commented:
Thanks @perfinion.  The problem with hardcoded paths at all is that we don't really use them for conda packages.  $PREFIX points to some path that looks like /usr or /usr/local (with include, bin, and lib subdirs), but it is not a fixed path.  For example, during a package build, it will be something like  

``/home/msarahan/miniconda3/conda-bld/tensorflow_(timestamp)/_h_env_(pad)``

with (pad) representing enough placeholder characters to get us to a 255 character padding for any fixed paths that get baked into binaries.  That's essential to relocatability.

Not having dynamic paths is a pretty big non-starter for us.

Adding --action_env=PREFIX is definitely something I knew we'd need to use - I'm just curious about how to incorporate that into the BUILD files.

To put headers and libs onto the search paths, I first symlinked my $PREFIX var in my build script to a location within the Tensforflow third_party folder, and added that folder to the include and link paths:

```
licenses(["notice"])  # BSD/MIT-like license

filegroup(
    name = "LICENSE",
    visibility = ["//visibility:public"],
)

cc_library(
    name = "png",
    copts = ["-Ithird_party/systemlibs/include"],
    linkopts = ["-lpng", "-Lthirdparty/systemlibs/lib"],
    visibility = ["//visibility:public"],
)
```

That worked OK, seemingly.  Next I was hung up on the hard-coded paths as I mentioned before.  I began to try something a little different, based on what I found at https://stackoverflow.com/questions/43928653/call-llvm-config-prefix-and-use-it-in-a-build-rule/43936381#43936381.

I removed the BUILD files completely, and instead tried to generate them more dynamically:

```
def _curl_impl(repository_ctx):
  prefix_path = repository_ctx.os.getenv("PREFIX", "/usr")
  repository_ctx.symlink(prefix_path, "external_prefix")
  repository_ctx.file("curl.BUILD", """
licenses(["notice"])  # MIT/X derivative license

filegroup(
    name = "COPYING",
    visibility = ["//visibility:public"],
)

cc_library(
    name = "curl",
    copts = ["-Iexternal_prefix/include"],
    linkopts = ["-lcurl", "-Lexternal_prefix/lib"],
    visibility = ["//visibility:public"],
)
""")
```

I have not gotten this working yet, but I think it might work.  It would still require --action_env=PREFIX, but that's fine with me.  Any thoughts on the viability of this approach?  It need not be part of this PR, if it is better to get this initial work in and then adapt.

perfinion commented:
@msarahan for cython specifically, it used to prefix it with $PYTHON_BIN_PATH which made things not work on gentoo since we have a special thing to select which python version is the default. that has been reverted so looks like its really easy now.

You dont need to bother with any of the generating stuff, the cmd= and copts and linkopts can all do make and bash variable expansion.
https://docs.bazel.build/versions/master/be/make-variables.html#make-var-substitution
https://docs.bazel.build/versions/master/be/common-definitions.html#sh-tokenization

You should be able to have just cython.BUILD with this:
```
genrule(
    name = "lncython",
    outs = ["cython.py"],
    cmd = "ln -s $$(which cython) $@",
)

sh_binary(
    name = "cython_binary",
    srcs = ["cython.py"],
    visibility = ["//visibility:public"],
)
```

Alternatively, in .bazelrc set `build --define PREFIX=/usr` then use `cmd = "ln -s $(PREFIX)/bin/cython"`
The problem is if PREFIX is not defined it fails, I havent figured out yet how to make it default to a blank value.

putting an `echo $${PREFIX}` inside cmd= does not work either but I think it should. That would be the best since we could use bash defaulting to do eg: "${PREFIX:-/usr}/bin/cython"

In copts and linkopts you can probably do "-I$(PREFIX)/include".

msarahan commented:
Seems very close to working, but I'm confused on an error with the external grpc.  I have put up my patch (I am applying it to 1.9.0rc1 right now): https://gist.github.com/msarahan/e38ccd45521617356a099d50b172f2b8

The error is:

```
Analyzing: target //tensorflow/tools/pip_package:build_pip_package (149 packages loaded)
Analyzing: target //tensorflow/tools/pip_package:build_pip_package (149 packages loaded)
ERROR: /Users/msarahan/miniconda3/conda-bld/tensorflow-base_1530276141984/work/tensorflow/core/distributed_runtime/rpc/BUILD:97:1: no such package '@grpc//': tf_http_archive rule //external:grpc must create a directory and referenced by '//tensorflow/core/distributed_runtime/rpc:grpc_channel'
ERROR: Analysis of target '//tensorflow/tools/pip_package:build_pip_package' failed; build aborted: no such package '@grpc//': tf_http_archive rule //external:grpc must create a directory
INFO: Elapsed time: 9.950s
```

My build script calls bazel with: 

```
     bazel ${BAZEL_OPTS} build \
           --verbose_failures \
           --config=opt \
           --action_env TF_SYSTEM_LIBS="${TF_SYSTEM_LIBS}" \
           --define PREFIX="${PREFIX}" \
           //tensorflow/tools/pip_package:build_pip_package
```

Before I was passing in --define as you recommended, it was dying in earlier externals, such as png.  That gives me confidence in the overall approach, but it seems I have something wrong.  If I remove grpc from the list of externals, it gets further, but dies on missing the license from LMDB.  It should not be necessary to list actual files in the external stuff, right?

perfinion commented:
@msarahan I rebased everything and also added the $(which foo) stuff

about your grpc error: its probably cuz you need to use double $$'s here:
cmd = "ln -s $(which grpc_python_plugin) $@",

I added those parts tho already so hopefully it should work if you just add the linkopts and includes.
I wonder if its easier for now to just add the linkopts and copts globally in the bazelrc?

perfinion commented:
I think I'm satisfied now with this patchset so far. It should be ready for review then merging when the workflow rejig is done.

I've pushed the older version with the gentoo _rc1-r2 ebuild. The _rc2 ebuild with the latest patches will be pushed to gentoo once some of the deps are bumped too.

martinwicke commented:
Can you fix the buildifier failure?

angersson commented:
@perfinion Thanks for keeping up with updating this!

Would you mind commenting with some commands to demonstrate how to test this on Ubuntu? It'll double as documentation for later. I'm far from an expert on builds, and as a user I wouldn't know how to go from 'jpeg is one of the packages that can come unbundled' to 'I need to `apt install libjpeg-dev` and then provide `--build=...`.

perfinion commented:
@martinwicke I rebased and fixed up the style change buildifier wanted. can you re-trigger the tests? :)

perfinion commented:
@angersson I'm not 100% sure on the package names on ubuntu but give this a shot.

To build tensorflow using the system libs for jpeg and zlib do:
```
$ apt-get install libjpeg-turbo8 libjpeg-tubo8-dev zlib1g zlib1g-dev
$ bazel clean --expunge
$ ./configure # answer everything normally
$ echo 'build --action_env TF_SYSTEM_LIBS="jpeg,zlib"' >> .tf_configure.bazelrc
$ bazel build --config=opt //tensorflow/tools/pip_package:build_pip_package \
//tensorflow:libtensorflow_framework.so \
//tensorflow:libtensorflow.so \
//tensorflow:libtensorflow_cc.so
```

----

To decide which libs can be unbundled, 
- Look in the `tensorflow/workspace.bzl` file.
- The entries that have a `system_build_file` should work.
- Add the `name` to the `TF_SYSTEM_LIBS` line.
- Make sure the library and header files for that package are installed in the system. The package names will differ per-distro (eg some might need to install the -dev package as well)
- If ubuntu has new enough versions of all the libs, unbundling all of them so far should work in theory but I have only tested on Gentoo so far.
- PREFIX installs are not easily possible yet. I wanted the basic way working first. I also need to find a good way to default the variable if not set. I will probably add stuff to ./configure to streamline the whole process but for now the beginning steps are still manual.

----

To unbundle a new library:
- Look in the bundled BUILD file to see which targets must be exported so tensorflow can use
- Write a new foo.BUILD file with the same targets as cc_library but linking to the system lib.
- Add the foo.BUILD file in third_party/systemlibs/foo.BUILD
- Add the `system_build_file=` key to `tensorflow/workspace.bzl` pointing to foo.BUILD
- Add the new entry in `VALID_LIBS` in `third_party/systemlibs/syslibs_configure.bzl`
- Do the above building steps to test, adjust foo.BUILD, repeat.

Using the basic cc_library rules is a bit cumbersome but it does work. I'd love for bazel to have some special rules that make system stuff easier. I realize it goes against bazels hermetic philosophy but I think its one hurdle that stops bazel from being useful in a lot of other places. The system rules could be gated behind a flag perhaps. And I can still see mostly-hermetic builds with bazel as being positive because it has all the extra header validation and dep checking that other systems don't have.

I should probably write more detail about this on my blog too.

jart commented:
Would you be able to put something like this in your ebuild file instead?

```sh
sed -i '
    /tf_http_archive(/!b
    :a
    N
    /)$/!ba
    /name = "curl"/bb
    /name = "gif_archive"/bb
    /name = "jpeg"/bb
    /name = "nasm"/bb
    /name = "png_archive"/bb
    /name = "snappy"/bb
    /name = "zlib_archive"/bb
    b
    :b
    s/\( *\).*\(name = "[^"]*"\).*/\1native.local_repository(\2, path=".system")/
  ' tensorflow/workspace.bzl

mkdir -p .system
touch .system/WORKSPACE
ln -sf $(command -v curl) .system/curl_bin
ln -sf $(command -v nasm) .system/nasm
for f in LICENSE LICENSE.txt LICENSE.md COPYING COPYING.txt; do
  echo 'dynamically linked via system' >.system/$f
done

cat >.system/BUILD <<EOF
package(default_visibility=["//visibility:public"])
exports_files(glob(["*"]))
cc_library(name="curl", linkopts=["-lcurl"])
cc_library(name="gif", linkopts=["-lgif"])
cc_library(name="jpeg", linkopts=["-ljpeg"])
cc_library(name="png", linkopts=["-lpng"])
cc_library(name="zlib", linkopts=["-lz"])
cc_library(name="snappy", linkopts=["-lsnappy"])
EOF
```

```sh
$ cd tensorflow
$ run-script-above.sh
$ bazel build //tensorflow:libtensorflow_framework.so
$ ldd bazel-bin/tensorflow/libtensorflow_framework.so | awk '{print $1}'
linux-vdso.so.1
libdl.so.2
libm.so.6
libsnappy.so.1
libz.so.1
libpthread.so.0
libgif.so.7
libjpeg.so.62
libpng16.so.16
libstdc++.so.6
libgcc_s.so.1
libc.so.6
/lib64/ld-linux-x86-64.so.2
```

Bazel seems to allow this hackishness, even with the sandboxing features enabled.

perfinion commented:
@jart I don't follow. That appears to work, yeah, but I don't get why?

Do you mean put it in the gentoo ebuild instead of upstreaming things? Then other distros don't benefit and everyone packaging TF has to duplicate the work.

Or you mean to put all the rules in one big workspace instead of separate? That works but makes toggling each dep separately more complicated. Maybe not everyone can unbundle everything so all--or-none seems less flexible.

jart reviewed (dismissed):
Would it be possible to integrate with pkg-config, pkgconf, etc. so it can generalize more across distros?

perfinion commented:
@jart pkg-config isn't doable nicely yet with bazel. I'm waiting on https://github.com/bazelbuild/bazel/issues/5188 and https://github.com/bazelbuild/bazel/issues/5186

I was going to expand the rules and iterate in later PRs once this patch set is merged. This one is already huge and gets merge conflicts all over the place. I have to keep rebasing the core parts every few days. Once the main parts are in, future changes will be much easier and less invasive.

perfinion committed:
third_party/repo: add TF_SYSTEM_LIBS to environ so bazel knows to refresh

Signed-off-by: Jason Zaman <jason@perfinion.com>

perfinion commented:
Also, just to be clear, this patchset does already work on other distros than just Gentoo. The lib names and linker opts are standardized.

The only thing that does not work right now is searching a different prefix other than the base compiler paths (@msarahan's usecase). But a regular .deb or .rpm would work. That one is what is more complicated by bazel's rules. I'll fix those in a later series, either with new bazel or add things to ./configure.

I'd rather get this merged before the 1.10 branch happens so other distros can get started packaging this too.

jart commented:
Which distros?

martinwicke commented:
@jart your solution is nice because it avoids more ifs in the BUILD. 

Is there a way to do this type of thing without running a separate script? Maybe by putting the if in a single place, say tf_lib_dep? tf_http_archive_switchable_dep? 

The goal of this work is for people to be able to build TF with system lib, and I don't think they should have to use sed to do it. We *could* put this thing in ./configure and ask for the libs to swap out, but that breaks another requirement, that dependent projects should be able to build TF by passing things to bazel and just run bazel.

perfinion commented:
I've tested ubuntu 18.04 too.
```
$ sudo apt-get install libjpeg-turbo8 libjpeg-turbo8-dev \
zlib1g zlib1g-dev libsnappy1v5 \
libsnappy-dev libre2-4 libre2-dev
$ bazel build --action_env \
TF_SYSTEM_LIBS="com_googlesource_code_re2,jpeg,snappy,zlib_archive" \
//tensorflow:libtensorflow_framework.so \
//tensorflow:libtensorflow.so \
//tensorflow:libtensorflow_cc.so \
//tensorflow/tools/pip_package:build_pip_package
```
I'm sure unbundling more would work, but those are the ones I've tested.

martinwicke commented:
I'm in favor of merging this. If we can simplify it later, great. I have the following questions:

- What happens when we add a new dependency? 
- What happens when we update a dependency?
- Can we remove some of the redundancy? Bazel likes things to be explicit, but having both a file in systemlibs and having to list its name in a whitelist is jarring (to me).

perfinion commented:
@martinwicke 
I could remove the whitelist, I thought it was safer to have but have no strong feelings either way. Could also remove it in the future if it ends up being a headache frequently. I was thinking later I could add a TF_SYSTEM_LIBS="all" which would just enable the whole whitelist.

When you add a new dep it will be bundled like things were in the past to start with. If unbundling it works (ie it has a release etc) then I or someone else from a distro can put together the unbundling rules.

If you update a dep, in most cases nothing will be needed in the systemlibs side. The libs would have the same name so the compiler would find it just the same as before. Distro packages would need to keep track of the minimum versions of deps but that goes for any package. If the dep changes significantly then the BUILD file will need updating but should be much less than the non-systemlib BUILD files

jart commented:
One alternative you could consider is creating a WORKSPACE file in a subdirectory and not calling `tf_workspace()`. In that case, Bazel's default behavior is to unbundle and let you define dependencies yourself. It's the only way to be certain the the unbundling is complete.

If you partially unbundle, you'll likely want to run TF's tests after building. For example, not unbundling protobuf might get you into trouble. Yesterday I encountered an issue where `import tensorflow` fails if protoc / libproto / pyproto versions mismatch: https://github.com/google/protobuf/issues/4716. Another example is our libpng config assumes a specific version of zlib. More such subtleties likely await those who partially interchange components.

angerson reviewed (dismissed):
Thanks for the examples! I'm in favor of getting this in, even if it needs some more iteration, since it shouldn't affect anyone that's not using it.

martinwicke commented:
@perfinion can you check the buildifier and --nobuild errors?

perfinion commented:
@martinwicke @angersson I fixed the buildifier style fix and rebased. can you re-trigger the tests?

I think the nobuild error was unrelated, looks like this commit fixes it. The rebase should pick that up too.

```
commit b13b2c15d2e864270b28eb5c5d3ec9f53d3932a7
Author: Yu-Cheng Ling <ycling@google.com>
Date:   Fri Jul 13 08:38:45 2018

    Fix the broekn BUILD rule under lite/delegates/eager.

nobuild error:
ERROR: /tmpfs/src/github/tensorflow/tensorflow/contrib/lite/delegates/eager/BUILD:22:1: no such package 'testing/base/public': BUILD file not found on package path and referenced by '//tensorflow/contrib/lite/delegates/eager:util_test'
ERROR: Analysis of target '//tensorflow/contrib/lite/delegates/eager:util_test' failed; build aborted: no such package 'testing/base/public': BUILD file not found on package path
```

angersson commented:
The changes here need manual intervention to work with the new PR import process, which I am currently working on handling.

angersson commented:
Done and merged! Thanks!

dennisjenkins75 commented:
Yeah!  Thank you all for vastly improving the TensorFlow experience on Gentoo Linux.

meteorcloudy commented:
/cc @dslomov

timokau commented:
@perfinion do you observe a significant change in build time on gentoo when using  system libraries? I tried to adopt this for the nix package, but to my surprise the build time barely changed (around 1h on a 12 core machine, 3.5h on a 4 core).

perfinion commented:
@timokau Sorry, didnt see this earlier, yeah its a fair bit faster if you unbundle all of them. GRPC, protobuf and several others are decently big but you're right stuff like libpng are fairly small so are not a huge compile time. The biggest part I've noticed tho is the linking stage is a lot faster and uses a *lot* less RAM than when static linking. Also the python deps were really annoying unless unbundled.

Also feel free to email me (jason at perfinion dot com) if you have any issues with packaging, i'd love to help but I might not see this thread. :)

timokau commented:
Thank you for your feedback and the offer :)

We have since managed to [get the source build working](https://github.com/NixOS/nixpkgs/pull/64716) and made use of [most](https://github.com/NixOS/nixpkgs/blob/6ee979916492e68871eef1d3bcf07841d2c987df/pkgs/development/python-modules/tensorflow/default.nix#L156) of the decoupled dependencies. Bundled deps are not quite as painful with nix (as everything has its own FHS etc.), but it's still nice to be able to unbundle them. Thank you for your work on this!
